Add clone methods to styles

simonseyock opened the pull request:
This pull request adds clone methods to most style objects. The clone methods are supposed to clone all properties which support getting cloned.

simonseyock commented on src/ol/style/regularshape.js:
```diff
@@ -113,7 +113,13 @@ ol.style.RegularShape = function(options) {
+   * @private
+   * @type {ol.style.AtlasManager|undefined}
+   */
+  this.atlasManager_ = options.atlasManager;
```
I am saving the atlasManger to apply it to the cloned style, too. Is this desired?

simonseyock replied:
Instead there could be an optional atlasManager parameter for clone for shape styles.

simonseyock commented on test/spec/ol/style/circle.test.js:
```diff
@@ -76,6 +76,40 @@ describe('ol.style.Circle', function() {
+      expect(original.getScale()).to.be(clone.getScale());
+      expect(original.getSnapToPixel()).to.be(clone.getSnapToPixel());
+      expect(original.getStroke()).to.not.be(clone.getStroke());
+      expect(original.getStroke().getColor()).to.be(clone.getStroke().getColor());
```
These kind of test can only check the values which have getters. Private properties cant be tested, this is not a problem for the circle, but that would be a problem for the icon. Think of the color to taint it. Is there a way to check if the style is really rendered the same?

simonseyock commented on src/ol/style/style.js:
```diff
@@ -72,6 +72,22 @@ ol.style.Style = function(opt_options) {
+ */
+ol.style.Style.prototype.clone = function() {
+  return new ol.style.Style({
+    geometry: this.getGeometry(),
```
should the geometry get cloned?

marcjansen commented:
I would expect a deep clone. This would mean to clone everything that is cloneable (e.g. geometries) changing sth. In th clone should usually not affect the original and vice versa. There are certain properties where this will be hard/undoable (geometryfunction, e.g.)

simonseyock commented:
Sounds reasonable.

What i did not look into yet is a clone method for the AtlasManager. I dont think it is needed. What do you think?

marcjansen commented:
TBH I know next to nothing about the AtlasManager. @ahocevar hinted he might rework that anyway.

Arrays, like the colors, might also need special love.

ahocevar commented:
See #5775 for what I had in mind. It would mean to take anything renderer specific out of `ol.style.*` (mostly affects `ol.style.Image` and subclasses). The atlas manager would then become a WebGL renderer option.

ahocevar commented:
But what I said above should not affect this pull request. I don't think people would expect the atlas manager to be cloned, but @tsauerwein may have better advice here.

tsauerwein commented:
> What I said above should not affect this pull request. I don't think people would expect the atlas manager to be cloned, but @tsauerwein may have better advice here.

Yes, it makes sense not to clone the atlas manager. It is meant to be shared among styles.

simonseyock commented on test/spec/ol/style/fill.test.js:
```diff
@@ -0,0 +1,24 @@
+        color: '#319FD3'
+      });
+      var clone = original.clone();
+      expect(original.getColor()).to.be(clone.getColor());
```
this will be changed to: to.eql

simonseyock commented:
Hey, i added a clone functionality for the image property of the icon style. What do you think about it?

simonseyock commented on src/ol/style/icon.js:
```diff
@@ -179,17 +179,29 @@ ol.inherits(ol.style.Icon, ol.style.Image);
+      newImage.height = oldImage.height;
+      context.drawImage(oldImage, 0, 0);
+    }
+  }
```
image and canvas cloning

marcjansen replied:
Looks OK to me, yes.

marcjansen commented:
I will have a look tomorrow, OK?

simonseyock commented:
Sure :) Thanks in advance

marcjansen commented on src/ol/style/fill.js:
```diff
@@ -31,6 +31,18 @@ ol.style.Fill = function(opt_options) {
+ */
+ol.style.Fill.prototype.clone = function() {
+  return new ol.style.Fill({
+    color: (this.getColor() instanceof Array) ? this.getColor().slice(0) : this.getColor()
```
What about a utility method in the `ol.color` namespace?

```js
ol.color.clone(color) {
  var clone = color;
  if (Array.isArray(color)) {
    clone = Array.from(color);
  } else if (color instanceof Uint8Array) {
    clone = Uint8Array.from(color);
  } else if (color instanceof Uint8ClampedArray) {
    clone = Uint8ClampedArray.from(color);
  }
  // String, CanvasPattern or CanvasGradient
  // are left untouched and not cloned
  return clone;
}
```

Untested.

simonseyock replied:
Sounds good!

simonseyock replied:
if i would use slice anyways to shallow copy an array, i could use this simplified variant:

```
ol.color.clone = function(color) {
  if (color.slice) {
    return color.slice();
  } else {
    return color;
  }
};
```

tschaub replied:
I'm slightly curious why @marcjansen keeps suggesting `Array.from()` (and friends), but think @simonseyock understands we don't need it (`foo.slice()` creates a `Uint8ClampedArray` if `foo` is a `Uint8ClampedArray`).

simonseyock replied:
@marcjansen already agreed on using `.slice()`. i just wanted to ask about skipping all the instanceof checks.

tschaub replied:
I think checking for `slice` is fine (and quite a bit nicer than `if (typeof color === 'string' || Array.isArray(color) || color instanceof Uint8Array || color instanceof Uint8ClampedArray)`).

As a side note, it looks like `ol.color.asArray()` wouldn't work when called with a `Uint8Array` or `Uint8ClampedArray`.  I'll create a ticket for that.

tschaub replied:
I've proposed simplifying the `ol.Color` type in #5872 (to make the `ol.color` functions work with `ol.Color`).

marcjansen commented on src/ol/style/icon.js:
```diff
@@ -172,6 +174,47 @@ ol.inherits(ol.style.Icon, ol.style.Image);
+  var newImage;
+  if (this.iconImage_.getImageState() === ol.Image.State.LOADED) {
+    if (oldImage.tagName.toUpperCase() === 'IMG') {
+      newImage = /** @type {Image} */ (this.getImage(1).cloneNode(true));
```
Reuse `oldImage`?

simonseyock replied:
Yes, that is an option, i was not sure if an image would be altered after giving it to a style. It seems a bit unlikely to me. You agree on cloning the canvas, though?

marcjansen replied:
With reusing I meant replacing `this.getImage(1)` with the already existing `oldImage` so that the line reads:

```js
newImage = /** @type {Image} */ (newImage.cloneNode(true));
```

simonseyock replied:
Ah well, sure :)

marcjansen commented on src/ol/style/icon.js:
```diff
@@ -172,6 +174,47 @@ ol.inherits(ol.style.Icon, ol.style.Image);
+    }
+  }
+  return new ol.style.Icon({
+    anchor: this.anchor_.slice(0),
```
Better `Array.from` and hint that a polyfill might be needed? Unsure here. Wouldn't block a `slice(0)`.

tschaub replied:
Do we not expect `this.anchor_` to be an array?

I think `array.slice()` is a fine way to shallow clone an array.

marcjansen replied:
Fine for me.

marcjansen commented on src/ol/style/icon.js:
```diff
@@ -172,6 +174,47 @@ ol.inherits(ol.style.Icon, ol.style.Image);
+    anchorXUnits: this.anchorXUnits_,
+    anchorYUnits: this.anchorYUnits_,
+    crossOrigin: this.crossOrigin_,
+    color: this.color_ !== null ? this.color_.slice(0) : undefined,
```
Reuse the clone method suggested above

marcjansen commented on src/ol/style/regularshape.js:
```diff
@@ -140,6 +146,31 @@ ol.inherits(ol.style.RegularShape, ol.style.Image);
+    fill: this.getFill() ? this.getFill().clone() : undefined,
+    points: this.getPoints(),
+    radius: this.getRadius(),
+    radius2: this.getRadius2(),
```
According to the API docs there is also a `radius1` configuration for `RegularShape`. I cannot find it in the OpenLayers code right now, though… :astonished:

bartvde replied:
https://github.com/openlayers/ol3/blob/00905e813123036d5ed85899d931b787647759e8/src/ol/style/regularshape.js#L70:L71 ?

bartvde replied:
But radius1 is just an alias for radius

marcjansen replied:
Ahhh! Thanks Bart.

marcjansen commented on src/ol/style/stroke.js:
```diff
@@ -63,6 +63,23 @@ ol.style.Stroke = function(opt_options) {
+ */
+ol.style.Stroke.prototype.clone = function() {
+  return new ol.style.Stroke({
+    color: (this.getColor() instanceof Array) ? this.getColor().slice(0) : this.getColor(),
```
Reuse the `ol.color.clone` method suggested above

marcjansen commented on src/ol/style/style.js:
```diff
@@ -72,6 +72,26 @@ ol.style.Style = function(opt_options) {
+  }
+  return new ol.style.Style({
+    geometry: geometry,
+    fill: this.getFill() ? this.getFill().clone() : undefined,
```
What about `image`? Did you leave it out intentionally

simonseyock replied:
Uhh, that slipped, thanks! Later i will add a test for this, too.

marcjansen commented on src/ol/style/text.js:
```diff
@@ -96,6 +96,28 @@ ol.style.Text.DEFAULT_FILL_COLOR_ = '#333';
+  return new ol.style.Text({
+    font: this.getFont(),
+    rotation: this.getRotation(),
+    rotateWithView: this.getRotateWithView(),
```
The above two lines work but aren't mentoined on the API docs… hmmm.

simonseyock replied:
That was added recently, are you sure you checked the most up to date documentation?

marcjansen replied:
Not sure right now… might have been looking at sth. outdated :see_no_evil:.

marcjansen commented:
Very nice work so far, @simonseyock! I have added a couple of inline comments and suggestions, but this is in good shape already.

I would be pleased if we had a special test where on changes properties in the clone and the original is unaffected and vice-versa. I know you have this when you e.g. say 

```js
expect(original.getFill()).to.not.be(clone.getFill());
// or somewhere else
expect(original.color_).to.not.be(clone.color_);
```

…but I would appreciate the extra verbosity in a dedicated test.

Also the description of this PR should be a little bit more descriptive and up-to-date.

Once you have updated the pull request, just ping me and I will gladly have a final look. Thanks so far!

A question for @openlayers/ol3 : Can we use `Array.from` and friends? IE 11 would need a polyfill, Edge should be OK.

marcjansen commented on src/ol/style/icon.js:
```diff
@@ -172,6 +174,47 @@ ol.inherits(ol.style.Icon, ol.style.Image);
+      newImage = /** @type {HTMLCanvasElement} */ (document.createElement('canvas'));
+      var context = newImage.getContext('2d');
+      newImage.width = oldImage.width;
+      newImage.height = oldImage.height;
```
Can be replaced with `ol.dom.createCanvasContext2D(oldImage.width, oldImage.height)`

Otherwise I like it.

simonseyock replied:
yes, i need the HTMLCanvasElement

marcjansen commented:
API docs of the clone methods should also state exactly what isn't cloned, IMHO.

tschaub commented:
> A question for @openlayers/ol3 : Can we use Array.from and friends? IE 11 would need a polyfill, Edge should be OK.

I think we should avoid making people use a polyfill when there is no clear reason for us to use the functionality it provides.  In the case of `Array.from()`, I don't think we have any need for it (we aren't dealing with `Map`, `Set`, or other array-like objects.

marcjansen commented:
As stated earlier, fine for me. Go ahead with slice.

simonseyock commented on src/ol/style/fill.js:
```diff
@@ -31,13 +31,14 @@ ol.style.Fill = function(opt_options) {
+  var color = this.getColor();
   return new ol.style.Fill({
-    color: (this.getColor() instanceof Array) ? this.getColor().slice(0) : this.getColor()
+    color: (color && color.slice) ? color.slice() : color || undefined
```
I backed down from using an ol.color.clone method because sometimes the type to clone is {ol.Color|string} and sometimes it {ol.Color|ol.ColorLike} and it was easier this way. I could use ol.colorlike.isColorLike, though.

simonseyock commented on src/ol/style/fill.js:
```diff
@@ -31,6 +31,19 @@ ol.style.Fill = function(opt_options) {
+ol.style.Fill.prototype.clone = function() {
+  var color = this.getColor();
+  return new ol.style.Fill({
+    color: (color && color.slice) ? color.slice() : color || undefined
```
I backed down from using an ol.color.clone method because sometimes the type to clone is {ol.Color|string} and sometimes it {ol.Color|ol.ColorLike} and it was easier this way. I could use ol.colorlike.isColorLike, though.

simonseyock commented on src/ol/style/regularshape.js:
```diff
@@ -401,11 +431,12 @@ ol.style.RegularShape.prototype.draw_ = function(renderOptions, context, x, y) {
-  for (i = 0; i <= this.points_; i++) {
-    angle0 = i * 2 * Math.PI / this.points_ - Math.PI / 2 + this.angle_;
+  for (i = 0; i <= points; i++) {
+    angle0 = i * 2 * Math.PI / points - Math.PI / 2 + this.angle_;
```
i introduced a new variable here to store the doubled points, else the value would be doubled after each clone (and each call of draw_ and drawHitDetectionCanvas_).

simonseyock commented on test/spec/ol/style.test.js:
```diff
@@ -1,97 +0,0 @@
-    }).to.throwException();
-  });
-
-});
```
I moved the style.test.js file to the style/ folder. I don't know if this fits into this PR, i could open an other PR for this.

simonseyock commented:
Hey, i added tests for the remaining clone methods and fixed some issues related to the clone methods. I think this PR is ready for review. Would be nice if you could look into it again @marcjansen.

simonseyock commented on src/ol/style/regularshape.js:
```diff
@@ -401,11 +431,12 @@ ol.style.RegularShape.prototype.draw_ = function(renderOptions, context, x, y) {
+  var points = this.points_;
   if (this.radius2_ !== this.radius_) {
-    this.points_ = 2 * this.points_;
+    points = 2 * points;
```
i introduced a new variable here to store the doubled points, else the value would be doubled after each clone (and each call of draw_ and drawHitDetectionCanvas_).

simonseyock replied:
This will cause a change of behaviour of the getPoints method:
It will now always return the initially given number of points, regardless of radius2 was given to the constructor or not.

marcjansen replied:
If I understand that correct you could have done the following alternatively (in `clone`), right:

```js
var style = new ol.style.RegularShape({
  // …
  points: this.radius2_ !== this.radius_ ? this.getPoints() / 2 : this.getPoints()
  // …
});
```

This way we wouldn't need the above changes, right?

simonseyock replied:
Yes, that is true, but for me it felt like a bug that getPoints would return the double amount of points i had given to the constructor. This isn't documented, too.

marcjansen replied:
Still, I would like to keep the amount of both
* code in this PR, and
* the number of changes for users
minimal.

If I construct a regular polygon with two circles each having 5 points, the total number of points is 10, right? So the current behaviour makes some sense (to me at least).

For this PR I suggest to not change the semantics of that method. Feel free to open an issue where to discuss this. I would not make that change in this PR as it makes it more complicated than it has to be.

It is basically the same you did with the `Style` test file. I would have left it where it is, to keep the changes to review for this PR minimal. (It is OK that you did move it, no worries to change that back again.). An issue where that is discussed would have been enough, I guess.

Don't get me wrong, I **really** appreciate the time and effort you have put into this already. I honestly want to get this in rather sooner than later, but the less changes you introduce, the less controversial it is.

simonseyock replied:
Ok, i agree with you, it is better to seperate the concerns. I open another issue for this. I thought that about the moving of the file, too. Thanks for the review!

tschaub commented:
@simonseyock - can you please rebase this on master?  Depending on the name of your remote, this could look something like this:

```js
git pull --rebase openlayers master
git push --force cloneStyle
```

(It looks like you are not set up with the same email everywhere you push commits.  You could solve this by running `git config --global user.email "your@email.com"` everywhere you run git.)

simonseyock commented:
Sure.

simonseyock commented:
I accidentally comitted a .orig file of kdiff, so i had to force push again.

marcjansen commented:
This is really looking good from my point of view, @simonseyock. I have added one comment with regard to regular shapes and `getPoints()`. If my understanding is correct we come away without changing the meaning of the returned value from `getPoints`.

I am :+1: for merging (once the `getPoints` thing is settled).

@tschaub, do you want to weigh in?

simonseyock commented:
I restored the behaviour of ol.style.RegularShape#getPoints. (and it catched my eye that i was using eql and be a little inconsistent, but this does not change anything)

marcjansen commented:
Thanks for the work on this and the patience to get it in, @simonseyock. Great work.

If someone has additional comments on this, we can always do it in follow-up PRs.
